Competition/src/main/cpp/Drivetrain.cpp

```diff
@@ -7,15 +7,231 @@
 }
 
 void Drivetrain::Periodic() {
-  // Update odometry
+  m_odometry.Update(frc::Rotation2d(units::degree_t(GetHeading())), GetLeftDistance(), GetRightDistance());
+  frc::SmartDashboard::PutNumber("Heading", imu.GetAngle());
+}
+
+void Drivetrain::InitDrives(rev::CANSparkMax::IdleMode idleMode) {
+  m_leftDriveLead.RestoreFactoryDefaults();
+  m_leftDriveFollowA.RestoreFactoryDefaults();
+  m_leftDriveFollowB.RestoreFactoryDefaults();
+  m_rightDriveLead.RestoreFactoryDefaults();
+  m_rightDriveFollowA.RestoreFactoryDefaults();
+  m_rightDriveFollowB.RestoreFactoryDefaults();
+
+  m_leftDriveLead.SetIdleMode(idleMode);
+  m_leftDriveFollowA.SetIdleMode(idleMode);
+  m_leftDriveFollowB.SetIdleMode(idleMode);
+  m_rightDriveLead.SetIdleMode(idleMode);
+  m_rightDriveFollowA.SetIdleMode(idleMode);
+  m_rightDriveFollowB.SetIdleMode(idleMode);
+
+  m_leftDriveLead.Follow(rev::CANSparkMax::kFollowerDisabled, false);
+  m_rightDriveLead.Follow(rev::CANSparkMax::kFollowerDisabled, false);
+
+  m_leftDriveFollowA.Follow(m_leftDriveLead);
+  m_leftDriveFollowB.Follow(m_leftDriveLead);
+
+  m_rightDriveFollowA.Follow(m_rightDriveLead);
+  m_rightDriveFollowB.Follow(m_rightDriveLead);
+
+  m_leftPIDController.SetP(DriveConstants::kP);
+  m_leftPIDController.SetI(DriveConstants::kI);
+  m_leftPIDController.SetD(DriveConstants::kD);
+  m_leftPIDController.SetIZone(DriveConstants::kIz);
+  m_leftPIDController.SetFF(DriveConstants::kFF);
+  m_leftPIDController.SetOutputRange(DriveConstants::kMinOutput, DriveConstants::kMaxOutput);
+
+  m_rightPIDController.SetP(DriveConstants::kP);
+  m_rightPIDController.SetI(DriveConstants::kI);
+  m_rightPIDController.SetD(DriveConstants::kD);
+  m_rightPIDController.SetIZone(DriveConstants::kIz);
+  m_rightPIDController.SetFF(DriveConstants::kFF);
+  m_rightPIDController.SetOutputRange(DriveConstants::kMinOutput, DriveConstants::kMaxOutput);
+
+  m_rightPIDController.SetSmartMotionMaxVelocity(DriveConstants::kMaxVel);
+  m_rightPIDController.SetSmartMotionMinOutputVelocity(DriveConstants::kMinVel);
+  m_rightPIDController.SetSmartMotionMaxAccel(DriveConstants::kMaxAccel);
+  m_rightPIDController.SetSmartMotionAllowedClosedLoopError(DriveConstants::kAllError);
+  
+  m_leftPIDController.SetSmartMotionMaxVelocity(DriveConstants::kMaxVel);
+  m_leftPIDController.SetSmartMotionMinOutputVelocity(DriveConstants::kMinVel);
+  m_leftPIDController.SetSmartMotionMaxAccel(DriveConstants::kMaxAccel);
+  m_leftPIDController.SetSmartMotionAllowedClosedLoopError(DriveConstants::kAllError);
+
+  m_leftDriveLead.SetInverted(false);
+  m_rightDriveLead.SetInverted(true);
+}
+
+void Drivetrain::ResetEncoders() {
+  m_leftEncoder.SetPosition(0);
+  m_rightEncoder.SetPosition(0);
+}
+
+void Drivetrain::ResetOdometry(frc::Pose2d pose) {
+  ResetEncoders();
+  m_odometry.ResetPosition(pose, frc::Rotation2d(units::degree_t(GetHeading())));
+}
+void Drivetrain::ResetIMU() {
+  imu.Reset();
+}
+
+// rev::CANEncoder& GetLeftEncoder() {
```

Comment:
Do we still need this to get encoder positions?

Reply:
No we do not. I think we made the method initially to get the encoder object but not necessary now.